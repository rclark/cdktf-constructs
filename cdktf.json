{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "f29a938c-2349-4541-8a17-5d9a640ede6d",
  "sendCrashReports": "true",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {}
}
