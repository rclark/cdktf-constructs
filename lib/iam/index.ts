export { NewRoleConfig, ExistingRoleConfig, PrincipalType, Role } from './role';
export { ServiceRole, ServiceRoleConfig } from './service-role';
