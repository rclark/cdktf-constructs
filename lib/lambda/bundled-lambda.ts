import { Construct } from 'constructs';
import { Lambda, LambdaConfig } from './lambda';

export interface BundledLambdaConfig
  extends Omit<
    LambdaConfig,
    'filename' | 'imageUri' | 'imageConfig' | 'packageType'
  > {
  s3Key: string;
  s3Bucket: string;
  runtime: string;
  handler: string;
}

export class BundledLambda extends Lambda {
  constructor(scope: Construct, id: string, config: BundledLambdaConfig) {
    super(scope, id, config);
  }
}
