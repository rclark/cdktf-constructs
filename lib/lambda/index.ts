export { Lambda, LambdaConfig } from './lambda';
