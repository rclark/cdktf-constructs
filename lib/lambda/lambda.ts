import { Construct } from 'constructs';
import { ITerraformDependable } from 'cdktf';
import {
  LambdaFunction,
  LambdaFunctionConfig,
  LambdaFunctionVpcConfig,
} from '@cdktf/provider-aws/lib/lambda-function';
import {
  CloudwatchLogGroup,
  CloudwatchLogGroupConfig,
} from '@cdktf/provider-aws/lib/cloudwatch-log-group';
import {
  CloudwatchMetricAlarm,
  CloudwatchMetricAlarmConfig,
} from '@cdktf/provider-aws/lib/cloudwatch-metric-alarm';
import { IamRolePolicy } from '@cdktf/provider-aws/lib/iam-role-policy';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { DataAwsIamRole } from '@cdktf/provider-aws/lib/data-aws-iam-role';
import { DataAwsIamPolicyDocument } from '@cdktf/provider-aws/lib/data-aws-iam-policy-document';
import { DataAwsRegion } from '@cdktf/provider-aws/lib/data-aws-region';
import * as iam from 'iam-floyd';

import { Role, ServiceRole } from '../iam';
import { DataAwsSubnetIds } from '@cdktf/provider-aws/lib/data-aws-subnet-ids';
import { DataAwsSecurityGroups } from '@cdktf/provider-aws/lib/data-aws-security-groups';
import { DataAwsVpc } from '@cdktf/provider-aws/lib/data-aws-vpc';
import { BaseConstruct } from '../core';

export interface LambdaRoleConfig {
  arn?: string;
  policies?: iam.PolicyStatement[];
}

export interface LambdaConfig
  extends Omit<LambdaFunctionConfig, 'role' | 'vpcConfig'> {
  logGroup?: Omit<CloudwatchLogGroupConfig, 'name' | 'namePrefix'>;
  role?: LambdaRoleConfig;
  alarm?: LambdaAlarmConfig;
}

export enum MetricStatistic {
  COUNT = 'SampleCount',
  AVG = 'Average',
  SUM = 'Sum',
  MIN = 'Minimum',
  MAX = 'Maximum',
}

export interface LambdaAlarmConfig
  extends Partial<
    Omit<
      CloudwatchMetricAlarmConfig,
      | 'alarmName'
      | 'alarmDescription'
      | 'namespace'
      | 'dimensions'
      | 'metricName'
      | 'statistic'
      | 'unit'
      | 'extendedStatistic'
      | 'evaluateLowSampleCountPercentiles'
      | 'metricQuery'
      | 'thresholdMetricId'
    >
  > {
  statistic?: MetricStatistic;
}

export interface LambdaVpcConfig extends Partial<LambdaFunctionVpcConfig> {
  vpcId?: string;
}

export class Lambda extends BaseConstruct {
  public static functionDefaults: Partial<LambdaFunctionConfig> = {
    memorySize: 128,
    timeout: 300,
  };

  public static alarmDefaults: Partial<CloudwatchMetricAlarmConfig> = {
    alarmActions: [],
    period: 60,
    statistic: MetricStatistic.SUM,
    datapointsToAlarm: 1,
    threshold: 0,
    comparisonOperator: 'GreaterThanThreshold',
    treatMissingData: 'notBreaching',
  };

  public static logGroupDefaults: Partial<CloudwatchLogGroupConfig> = {
    retentionInDays: 14,
  };

  private static fnId = 'fn';
  private static logId = 'logs';
  private static alarmId = 'alarm';
  private static regionId = 'region';
  private static logDocId = 'log-doc';
  private static policyId = 'policy';
  private static roleId = 'role';
  private static vpcId = 'vpc';
  private static subnetId = 'subnets';
  private static sgId = 'sgs';

  constructor(scope: Construct, id: string, config: LambdaConfig) {
    super(scope, id);

    this.functionName = config.functionName;
    this.region = new DataAwsRegion(scope, Lambda.regionId);

    this.buildRole(config.role);

    this.logGroup = new CloudwatchLogGroup(
      this,
      Lambda.logId,
      this.logCfg(config)
    );

    this.lambda = new LambdaFunction(this, Lambda.fnId, this.lambdaCfg(config));

    this.alarm = new CloudwatchMetricAlarm(
      this,
      Lambda.alarmId,
      this.alarmCfg(config)
    );
  }

  public lambda: LambdaFunction;
  public alarm: CloudwatchMetricAlarm;
  public loggingPermissionsDoc!: DataAwsIamPolicyDocument;
  public loggingPolicy!: IamRolePolicy;
  public vpc?: DataAwsVpc;
  public subnets?: DataAwsSubnetIds;
  public sgs?: DataAwsSecurityGroups;

  private set logGroup(group: CloudwatchLogGroup) {
    this.loggingPermissionsDoc = new DataAwsIamPolicyDocument(
      this,
      Lambda.logDocId,
      {
        statement: [
          {
            effect: 'Allow',
            actions: ['logs:*'],
            resources: [group.arn],
          },
        ],
      }
    );

    this.loggingPolicy = new IamRolePolicy(this, Lambda.policyId, {
      name: `lambda-logging-${this.functionName}`,
      role: this.role.name,
      policy: this.loggingPermissionsDoc.json,
    });

    this._logGroup = group;
  }

  public get logGroup(): CloudwatchLogGroup {
    return this._logGroup;
  }

  private functionName: string;
  private region: DataAwsRegion;
  private _logGroup!: CloudwatchLogGroup;
  private _role!: Role;

  public get role(): DataAwsIamRole | IamRole {
    return this._role.role;
  }

  public inVPC(config?: LambdaVpcConfig): Lambda {
    let { subnetIds, securityGroupIds, vpcId } = config || {};

    if (!config) {
      this.vpc = new DataAwsVpc(this, Lambda.vpcId, {});
      vpcId = this.vpc.id;
    }

    if (vpcId) {
      this.subnets = new DataAwsSubnetIds(this, Lambda.subnetId, {
        vpcId: vpcId,
      });
      subnetIds = subnetIds || this.subnets.ids;

      this.sgs = new DataAwsSecurityGroups(this, Lambda.sgId, {
        filter: [{ name: 'vpc-id', values: [vpcId] }],
      });
      securityGroupIds = securityGroupIds || this.sgs.ids;
    }

    if (!subnetIds || !securityGroupIds) {
      throw new Error(
        'Must supply either no vpc config, a vpcId, or both subnetIds and securityGroupIds'
      );
    }

    this.lambda.putVpcConfig({ subnetIds, securityGroupIds });
    return this;
  }

  private buildRole(config: LambdaRoleConfig | undefined) {
    if (config && config.arn) {
      const existing = new Role(this, Lambda.roleId, {
        arn: config.arn,
        policies: config.policies,
      });
      this._role = existing;
      return;
    }

    const newRole = new ServiceRole(this, Lambda.roleId, {
      name: `lambda-role-${this.functionName}`,
      statement: config ? config.policies : [],
      services: ['lambda'],
    });
    this._role = newRole;
  }

  private lambdaCfg(config: LambdaConfig): LambdaFunctionConfig {
    let depends = [this.logGroup as ITerraformDependable];
    if (config.dependsOn) depends = depends.concat(config.dependsOn);

    const required: LambdaFunctionConfig = {
      dependsOn: depends,
      functionName: this.functionName,
      role: this.role.arn,
    };

    return Object.assign({}, Lambda.functionDefaults, config, required);
  }

  private alarmCfg(config: LambdaConfig): CloudwatchMetricAlarmConfig {
    const required: Partial<CloudwatchMetricAlarmConfig> = {
      alarmName: `lambda-errors-${this.functionName}-${this.region.name}`,
      alarmDescription: `Error alarm for the ${this.functionName} Lambda function`,
      namespace: 'AWS/Lambda',
      dimensions: { FunctionName: this.functionName },
      metricName: 'Errors',
    };

    const intermediate = Object.assign(
      {},
      Lambda.alarmDefaults,
      config.alarm,
      required
    );

    const evaluationPeriods = Math.ceil(
      this.lambdaCfg(config).timeout! / intermediate.period!
    );

    return Object.assign(
      { evaluationPeriods },
      intermediate
    ) as CloudwatchMetricAlarmConfig;
  }

  private logCfg(config: LambdaConfig): CloudwatchLogGroupConfig {
    const required: Partial<CloudwatchLogGroupConfig> = {
      name: `/aws/lambda/${this.functionName}`,
    };

    return Object.assign(
      {},
      Lambda.logGroupDefaults,
      config.logGroup,
      required
    );
  }

  removeResources(): void {
    this._role.removeResources();

    this.node.tryRemoveChild(Lambda.fnId);
    this.node.tryRemoveChild(Lambda.logId);
    this.node.tryRemoveChild(Lambda.alarmId);
    this.node.tryRemoveChild(Lambda.logDocId);
    this.node.tryRemoveChild(Lambda.vpcId);
    this.node.tryRemoveChild(Lambda.sgId);
    this.node.tryRemoveChild(Lambda.subnetId);
    this.node.tryRemoveChild(Lambda.regionId);
    this.node.tryRemoveChild(Lambda.policyId);
  }
}
