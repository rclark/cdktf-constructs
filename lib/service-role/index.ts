import { Construct } from 'constructs';
import { TerraformMetaArguments } from 'cdktf';
import { PrincipalType, Role } from '../role';
import * as iam from 'iam-floyd';
import { DataAwsPartition } from '@cdktf/provider-aws/lib/data-aws-partition';

export interface ServiceRoleConfig extends TerraformMetaArguments {
  name: string;
  statement?: iam.PolicyStatement[];
  services: string[];
}

export class ServiceRole extends Role {
  constructor(scope: Construct, id: string, config: ServiceRoleConfig) {
    const partition = new DataAwsPartition(scope, id, config);

    const avoidSuffix = [
      'apigateway',
      'autoscaling',
      'cloudformation',
      'codedeploy',
      'ecs-tasks',
      'elasticbeanstalk',
      'firehose',
      'iot',
      'lambda',
      'rds',
      'redshift',
      's3',
      'sms',
      'storagegateway',
      'swf',
    ];

    const principals = config.services.map((value: string): string => {
      const prefix = value.replace(/\.amazonaws.com(\..*)?$/, '');
      const suffix = avoidSuffix.includes(prefix)
        ? 'amazonaws.com'
        : partition.dnsSuffix;
      return `${prefix}.${suffix}`;
    });

    super(scope, id, {
      name: config.name,
      principals: [{ type: PrincipalType.SERVICE, identifiers: principals }],
    });
  }
}
