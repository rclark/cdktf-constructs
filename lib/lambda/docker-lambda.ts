import { Construct } from 'constructs';
import { Lambda, LambdaConfig } from './lambda';

export interface DockerLambdaConfig
  extends Omit<
    LambdaConfig,
    | 'filename'
    | 'handler'
    | 'imageUri'
    | 'packageType'
    | 'runtime'
    | 's3Key'
    | 's3Bucket'
    | 's3ObjectVersion'
    | 'snapStart'
    | 'sourceCodeHash'
  > {
  imageUri: string;
}

export class DockerLambda extends Lambda {
  constructor(scope: Construct, id: string, config: DockerLambdaConfig) {
    super(scope, id, config);

    this.lambda.imageUri = config.imageUri;
    this.lambda.packageType = 'Image';
  }
}
