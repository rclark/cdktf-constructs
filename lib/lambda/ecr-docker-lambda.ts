import { spawn } from 'child_process';
import { Construct } from 'constructs';
import { ECR } from '@aws-sdk/client-ecr';

import { DockerLambda, DockerLambdaConfig } from './docker-lambda';
import { PrivateRepository } from '../ecr';

export interface EcrDockerLambdaBuildConfig {
  dockerfile: string;
  context: string;
}

export interface EcrDockerLambdaConfig
  extends Omit<DockerLambdaConfig, 'imageUri'> {
  build: EcrDockerLambdaBuildConfig;
  gitsha?: string;
}

export class EcrDockerLambda extends DockerLambda {
  private repo: PrivateRepository;
  private static repoId = 'repo';
  private buildConfig: EcrDockerLambdaBuildConfig;
  private name: string;
  private gitsha?: string;

  constructor(scope: Construct, id: string, config: EcrDockerLambdaConfig) {
    const imageUri = '';
    super(scope, id, Object.assign({ imageUri }, config));

    this.buildConfig = config.build;
    this.name = config.functionName;
    this.gitsha = config.gitsha;

    this.repo = new PrivateRepository(this, EcrDockerLambda.repoId, {
      name: this.name,
    }).allowLambdaAccess();

    this.lambda.imageUri = `${this.repo.repository.repositoryUrl}:${config.gitsha}`;

    if (!config.gitsha) super.removeResources();
  }

  public async build(): Promise<void> {
    if (!this.gitsha) return;

    const ecr = new ECR({});
    const repos = await ecr.describeRepositories({
      repositoryNames: [this.name],
    });
    const uri = repos.repositories![0].repositoryUri!;
    const image = `${uri}:${this.gitsha}`;

    const auth = await ecr.getAuthorizationToken({});
    const token = Buffer.from(auth.authorizationData![0].authorizationToken!, 'base64').toString().split(':')[1];

    await run(
      'docker',
      'build',
      '-f',
      this.buildConfig.dockerfile,
      '-t',
      image,
      this.buildConfig.context
    );

    await run(
      'docker',
      'login',
      '--username', 'AWS',
      '--password', token,
      uri.split('/')[0]
    )

    await run('docker', 'push', image)
  }

  removeResources(): void {
    super.removeResources();
    this.node.tryRemoveChild(EcrDockerLambda.repoId);
  }
}

async function run(cmd: string, ...args: string[]): Promise<void> {
  return new Promise((resolve, reject) => {
    console.log(cmd, ...args);
    const p = spawn(cmd, args);

    p.stdout.pipe(process.stdout);
    p.stderr.pipe(process.stderr);

    p.on('error', reject);
    p.on('close', resolve);
  });
}
