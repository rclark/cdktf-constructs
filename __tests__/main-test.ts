// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import 'cdktf/lib/testing/adapters/jest'; // Load types for expect matchers
import { TerraformStack, Testing } from 'cdktf';
import { Lambda } from '../lib/lambda';
import {
  ExistingRoleConfig,
  NewRoleConfig,
  PrincipalType,
  Role,
} from '../lib/iam';
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { DockerLambda, DockerLambdaConfig } from '../lib/lambda/docker-lambda';
import { MetricStatistic } from '../lib/lambda/lambda';

const fakeArn = 'arn:aws:service:us-west-2:123456789012:entity';
const fakeSecurityGroup = 'sg-7ca1556d';
const fakeSubnet = 'subnet-362d0717';

describe('cdk-constructs', () => {
  describe('Lambda', () => {
    it('builds with minimal configuration', () => {
      const synth = Testing.synthScope((scope) => {
        new Lambda(scope, 'my-function', { functionName: 'my-function' });
      });
      expect(synth).toMatchSnapshot();
    });
  });

  describe('DockerLambda', () => {
    describe('with minimal configuration', () => {
      const cfg: DockerLambdaConfig = {
        functionName: 'my-function',
        imageUri: 'public.ecr.aws/lambda/provided:al2',
      };

      it('builds', () => {
        const synth = Testing.synthScope((scope) => {
          new DockerLambda(scope, 'id', cfg);
        });
        expect(synth).toMatchSnapshot();
      });

      it('makes valid terraform', () => {
        const app = Testing.app();
        const stack = new TerraformStack(app, 'test');
        new AwsProvider(stack, 'aws', { region: 'us-west-2' });
        new DockerLambda(stack, 'id', cfg);

        expect(Testing.fullSynth(stack)).toBeValidTerraform();
      });
    });

    describe('with maximal configuration', () => {
      const cfg: DockerLambdaConfig = {
        functionName: 'my-function',
        imageUri: 'public.ecr.aws/lambda/provided:al2',
        architectures: ['x86_64'],
        codeSigningConfigArn: fakeArn,
        deadLetterConfig: {
          targetArn: fakeArn,
        },
        description: 'described',
        environment: { variables: { KEY: 'value' } },
        ephemeralStorage: { size: 512 },
        fileSystemConfig: { arn: fakeArn, localMountPath: '/mnt/something' },
        imageConfig: {
          command: ['foo'],
          entryPoint: ['bar'],
          workingDirectory: 'baz',
        },
        kmsKeyArn: fakeArn,
        layers: [fakeArn],
        memorySize: 512,
        publish: false,
        reservedConcurrentExecutions: -1,
        replaceSecurityGroupsOnDestroy: false,
        replacementSecurityGroupIds: [fakeSecurityGroup],
        skipDestroy: false,
        tags: { tag: 'value' },
        timeout: 3,
        tracingConfig: { mode: 'Active' },
        vpcConfig: {
          securityGroupIds: [fakeSecurityGroup],
          subnetIds: [fakeSubnet],
        },
        logGroup: {
          skipDestroy: false,
          retentionInDays: 7,
          kmsKeyId: fakeArn,
          tags: { tag: 'value' },
        },
        alarm: {
          comparisonOperator: 'GreaterThanOrEqualToThreshold',
          evaluationPeriods: 2,
          period: 120,
          statistic: MetricStatistic.AVG,
          threshold: 10,
          actionsEnabled: true,
          alarmActions: [fakeArn],
          datapointsToAlarm: 1,
          insufficientDataActions: [fakeArn],
          okActions: [fakeArn],
          treatMissingData: 'missing',
          tags: { tag: 'value' },
        },
      };

      it('builds', () => {
        const synth = Testing.synthScope((scope) => {
          new DockerLambda(scope, 'id', cfg);
        });
        expect(synth).toMatchSnapshot();
      });

      it('makes valid terraform', () => {
        const app = Testing.app();
        const stack = new TerraformStack(app, 'test');
        new AwsProvider(stack, 'aws', { region: 'us-west-2' });
        new DockerLambda(stack, 'id', cfg);

        expect(Testing.fullSynth(stack)).toBeValidTerraform();
      });
    });
  });

  describe('Role', () => {
    describe('with minimal configuration', () => {
      describe('a new role', () => {
        const cfg: NewRoleConfig = {
          name: 'new-role',
          principals: [
            {
              type: PrincipalType.AWS,
              identifiers: ['arn:aws:iam::1234567890:user/chuck'],
            },
          ],
        };

        it('builds', () => {
          const synth = Testing.synthScope((scope) => {
            new Role(scope, 'my-role', cfg);
          });
          expect(synth).toMatchSnapshot();
        });

        it('makes valid terraform', () => {
          const app = Testing.app();
          const stack = new TerraformStack(app, 'test');
          new AwsProvider(stack, 'aws', { region: 'us-west-2' });
          new Role(stack, 'my-role', cfg);

          expect(Testing.fullSynth(stack)).toBeValidTerraform();
        });
      });

      describe('an existing role', () => {
        const cfg: ExistingRoleConfig = {
          arn: 'arn:aws:iam::1234567890:role/developer',
        };

        it('builds', () => {
          const synth = Testing.synthScope((scope) => {
            new Role(scope, 'my-role', cfg);
          });
          expect(synth).toMatchSnapshot();
        });

        it('makes valid terraform', () => {
          const app = Testing.app();
          const stack = new TerraformStack(app, 'test');
          new AwsProvider(stack, 'aws', { region: 'us-west-2' });
          new Role(stack, 'my-role', cfg);

          expect(Testing.fullSynth(stack)).toBeValidTerraform();
        });
      });
    });
  });

  //   it("Tests a combination of resources", () => {
  //     expect(
  //       Testing.synthScope((stack) => {
  //         new TestDataSource(stack, "test-data-source", {
  //           name: "foo",
  //         });

  //         new TestResource(stack, "test-resource", {
  //           name: "bar",
  //         });
  //       })
  //     ).toMatchInlineSnapshot();
  //   });
  // });

  // describe("Checking validity", () => {
  //   it("check if the produced terraform configuration is valid", () => {
  //     const app = Testing.app();
  //     const stack = new TerraformStack(app, "test");

  //     new TestDataSource(stack, "test-data-source", {
  //       name: "foo",
  //     });

  //     new TestResource(stack, "test-resource", {
  //       name: "bar",
  //     });
  //     expect(Testing.fullSynth(app)).toBeValidTerraform();
  //   });

  //   it("check if this can be planned", () => {
  //     const app = Testing.app();
  //     const stack = new TerraformStack(app, "test");

  //     new TestDataSource(stack, "test-data-source", {
  //       name: "foo",
  //     });

  //     new TestResource(stack, "test-resource", {
  //       name: "bar",
  //     });
  //     expect(Testing.fullSynth(app)).toPlanSuccessfully();
  //   });
  // });
});
