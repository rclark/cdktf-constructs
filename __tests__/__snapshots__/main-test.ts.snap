// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-constructs DockerLambda with maximal configuration builds 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "id_my-function-log-permissions_8C1AFB0E": {
        "statement": [
          {
            "actions": [
              "logs:*"
            ],
            "effect": "Allow",
            "resources": [
              "\${aws_cloudwatch_log_group.id_id-logs_E7E8A5B1.arn}"
            ]
          }
        ]
      },
      "id_my-function-role_my-function-role-assume_3B36031C": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "lambda.amazonaws.com"
                ],
                "type": "Service"
              }
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "id_my-function-role-partition_C7AF5BB8": {
      }
    },
    "aws_region": {
      "id-region": {
      }
    }
  },
  "resource": {
    "aws_cloudwatch_log_group": {
      "id_id-logs_E7E8A5B1": {
        "kms_key_id": "arn:aws:service:us-west-2:123456789012:entity",
        "name": "/aws/lambda/my-function",
        "retention_in_days": 7,
        "skip_destroy": false,
        "tags": {
          "tag": "value"
        }
      }
    },
    "aws_cloudwatch_metric_alarm": {
      "id_id-alarm_EB9777B2": {
        "actions_enabled": true,
        "alarm_actions": [
          "arn:aws:service:us-west-2:123456789012:entity"
        ],
        "alarm_description": "Error alarm for the my-function Lambda function",
        "alarm_name": "lambda-errors-my-function-\${data.aws_region.id-region.name}",
        "comparison_operator": "GreaterThanOrEqualToThreshold",
        "datapoints_to_alarm": 1,
        "dimensions": {
          "FunctionName": "my-function"
        },
        "evaluation_periods": 2,
        "insufficient_data_actions": [
          "arn:aws:service:us-west-2:123456789012:entity"
        ],
        "metric_name": "Errors",
        "namespace": "AWS/Lambda",
        "ok_actions": [
          "arn:aws:service:us-west-2:123456789012:entity"
        ],
        "period": 120,
        "statistic": "Average",
        "tags": {
          "tag": "value"
        },
        "threshold": 10,
        "treat_missing_data": "missing"
      }
    },
    "aws_iam_role": {
      "id_my-function-role_my-function-role-new-role_2EB66BDC": {
        "assume_role_policy": "\${data.aws_iam_policy_document.id_my-function-role_my-function-role-assume_3B36031C.json}",
        "name": "lambda-role-my-function"
      }
    },
    "aws_iam_role_policy": {
      "id_my-function-log-policy_EDED9500": {
        "name": "lambda-logging-my-function",
        "policy": "\${data.aws_iam_policy_document.id_my-function-log-permissions_8C1AFB0E.json}",
        "role": "\${aws_iam_role.id_my-function-role_my-function-role-new-role_2EB66BDC.arn}"
      }
    },
    "aws_lambda_function": {
      "id_6B6273EC": {
        "architectures": [
          "x86_64"
        ],
        "code_signing_config_arn": "arn:aws:service:us-west-2:123456789012:entity",
        "dead_letter_config": {
          "target_arn": "arn:aws:service:us-west-2:123456789012:entity"
        },
        "depends_on": [
          "aws_cloudwatch_log_group.id_id-logs_E7E8A5B1"
        ],
        "description": "described",
        "environment": {
          "variables": {
            "KEY": "value"
          }
        },
        "ephemeral_storage": {
          "size": 512
        },
        "file_system_config": {
          "arn": "arn:aws:service:us-west-2:123456789012:entity",
          "local_mount_path": "/mnt/something"
        },
        "function_name": "my-function",
        "image_config": {
          "command": [
            "foo"
          ],
          "entry_point": [
            "bar"
          ],
          "working_directory": "baz"
        },
        "image_uri": "public.ecr.aws/lambda/provided:al2",
        "kms_key_arn": "arn:aws:service:us-west-2:123456789012:entity",
        "layers": [
          "arn:aws:service:us-west-2:123456789012:entity"
        ],
        "memory_size": 512,
        "package_type": "Image",
        "publish": false,
        "replace_security_groups_on_destroy": false,
        "replacement_security_group_ids": [
          "sg-7ca1556d"
        ],
        "reserved_concurrent_executions": -1,
        "role": "\${aws_iam_role.id_my-function-role_my-function-role-new-role_2EB66BDC.arn}",
        "skip_destroy": false,
        "tags": {
          "tag": "value"
        },
        "timeout": 3,
        "tracing_config": {
          "mode": "Active"
        },
        "vpc_config": {
          "security_group_ids": [
            "sg-7ca1556d"
          ],
          "subnet_ids": [
            "subnet-362d0717"
          ]
        }
      }
    }
  }
}"
`;

exports[`cdk-constructs DockerLambda with minimal configuration builds 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "id_my-function-log-permissions_8C1AFB0E": {
        "statement": [
          {
            "actions": [
              "logs:*"
            ],
            "effect": "Allow",
            "resources": [
              "\${aws_cloudwatch_log_group.id_id-logs_E7E8A5B1.arn}"
            ]
          }
        ]
      },
      "id_my-function-role_my-function-role-assume_3B36031C": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "lambda.amazonaws.com"
                ],
                "type": "Service"
              }
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "id_my-function-role-partition_C7AF5BB8": {
      }
    },
    "aws_region": {
      "id-region": {
      }
    }
  },
  "resource": {
    "aws_cloudwatch_log_group": {
      "id_id-logs_E7E8A5B1": {
        "name": "/aws/lambda/my-function",
        "retention_in_days": 14
      }
    },
    "aws_cloudwatch_metric_alarm": {
      "id_id-alarm_EB9777B2": {
        "alarm_actions": [
        ],
        "alarm_description": "Error alarm for the my-function Lambda function",
        "alarm_name": "lambda-errors-my-function-\${data.aws_region.id-region.name}",
        "comparison_operator": "GreaterThanThreshold",
        "datapoints_to_alarm": 1,
        "dimensions": {
          "FunctionName": "my-function"
        },
        "evaluation_periods": 5,
        "metric_name": "Errors",
        "namespace": "AWS/Lambda",
        "period": 60,
        "statistic": "Sum",
        "threshold": 0,
        "treat_missing_data": "notBreaching"
      }
    },
    "aws_iam_role": {
      "id_my-function-role_my-function-role-new-role_2EB66BDC": {
        "assume_role_policy": "\${data.aws_iam_policy_document.id_my-function-role_my-function-role-assume_3B36031C.json}",
        "name": "lambda-role-my-function"
      }
    },
    "aws_iam_role_policy": {
      "id_my-function-log-policy_EDED9500": {
        "name": "lambda-logging-my-function",
        "policy": "\${data.aws_iam_policy_document.id_my-function-log-permissions_8C1AFB0E.json}",
        "role": "\${aws_iam_role.id_my-function-role_my-function-role-new-role_2EB66BDC.arn}"
      }
    },
    "aws_lambda_function": {
      "id_6B6273EC": {
        "depends_on": [
          "aws_cloudwatch_log_group.id_id-logs_E7E8A5B1"
        ],
        "function_name": "my-function",
        "image_uri": "public.ecr.aws/lambda/provided:al2",
        "memory_size": 128,
        "package_type": "Image",
        "role": "\${aws_iam_role.id_my-function-role_my-function-role-new-role_2EB66BDC.arn}",
        "timeout": 300
      }
    }
  }
}"
`;

exports[`cdk-constructs Lambda builds with minimal configuration 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "my-function_my-function-log-permissions_1D3E84D7": {
        "statement": [
          {
            "actions": [
              "logs:*"
            ],
            "effect": "Allow",
            "resources": [
              "\${aws_cloudwatch_log_group.my-function_my-function-logs_A4870A22.arn}"
            ]
          }
        ]
      },
      "my-function_my-function-role_my-function-role-assume_4F65D8FD": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "lambda.amazonaws.com"
                ],
                "type": "Service"
              }
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "my-function_my-function-role-partition_7C5CD2BD": {
      }
    },
    "aws_region": {
      "my-function-region": {
      }
    }
  },
  "resource": {
    "aws_cloudwatch_log_group": {
      "my-function_my-function-logs_A4870A22": {
        "name": "/aws/lambda/my-function",
        "retention_in_days": 14
      }
    },
    "aws_cloudwatch_metric_alarm": {
      "my-function_my-function-alarm_EF5A6BA3": {
        "alarm_actions": [
        ],
        "alarm_description": "Error alarm for the my-function Lambda function",
        "alarm_name": "lambda-errors-my-function-\${data.aws_region.my-function-region.name}",
        "comparison_operator": "GreaterThanThreshold",
        "datapoints_to_alarm": 1,
        "dimensions": {
          "FunctionName": "my-function"
        },
        "evaluation_periods": 5,
        "metric_name": "Errors",
        "namespace": "AWS/Lambda",
        "period": 60,
        "statistic": "Sum",
        "threshold": 0,
        "treat_missing_data": "notBreaching"
      }
    },
    "aws_iam_role": {
      "my-function_my-function-role_my-function-role-new-role_EF570E59": {
        "assume_role_policy": "\${data.aws_iam_policy_document.my-function_my-function-role_my-function-role-assume_4F65D8FD.json}",
        "name": "lambda-role-my-function"
      }
    },
    "aws_iam_role_policy": {
      "my-function_my-function-log-policy_78428521": {
        "name": "lambda-logging-my-function",
        "policy": "\${data.aws_iam_policy_document.my-function_my-function-log-permissions_1D3E84D7.json}",
        "role": "\${aws_iam_role.my-function_my-function-role_my-function-role-new-role_EF570E59.arn}"
      }
    },
    "aws_lambda_function": {
      "my-function_0A1E7AB6": {
        "depends_on": [
          "aws_cloudwatch_log_group.my-function_my-function-logs_A4870A22"
        ],
        "function_name": "my-function",
        "memory_size": 128,
        "role": "\${aws_iam_role.my-function_my-function-role_my-function-role-new-role_EF570E59.arn}",
        "timeout": 300
      }
    }
  }
}"
`;

exports[`cdk-constructs Role with minimal configuration a new role builds 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "my-role_my-role-assume_AC40947D": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "arn:aws:iam::1234567890:user/chuck"
                ],
                "type": "AWS"
              }
            ]
          }
        ]
      }
    }
  },
  "resource": {
    "aws_iam_role": {
      "my-role_my-role-new-role_137CD40C": {
        "assume_role_policy": "\${data.aws_iam_policy_document.my-role_my-role-assume_AC40947D.json}",
        "name": "new-role"
      }
    }
  }
}"
`;

exports[`cdk-constructs Role with minimal configuration an existing role builds 1`] = `
"{
  "data": {
    "aws_iam_role": {
      "my-role_my-role-external-role_747D99F6": {
        "name": "developer"
      }
    }
  }
}"
`;
